orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc

```diff
@@ -0,0 +1,148 @@
+  return alloc;
+}
+
+Status GenerateOutputs(OpKernelContext* ctx, cudaStream_t steam,
+                       const TensorSeq* values, TensorSeq* updated_values,
+                       size_t number_of_values) {
+  // Return if the output edge is not fetched.
+  if (updated_values == nullptr) {
+    return Status::OK();
+  }
+
+  bool is_same_buffer = const_cast<TensorSeq*>(values) == updated_values;
+  if (!is_same_buffer) {
+    updated_values->SetType(values->DataType());
+    updated_values->Reserve(number_of_values);
+    for (size_t input_idx = 0; input_idx < number_of_values; ++input_idx) {
+      const Tensor& source_tensor = values->Get(input_idx);
+      std::unique_ptr<Tensor> target_tensor = Tensor::Create(source_tensor.DataType(),
+                                                             source_tensor.Shape(), GetAllocatorPtr(ctx));
+      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(target_tensor->MutableDataRaw(),
+                                           source_tensor.DataRaw(),
+                                           source_tensor.SizeInBytes(),
+                                           cudaMemcpyDeviceToDevice, steam));
+      updated_values->Add(std::move(*target_tensor));  // Add will check for type consistency
+    }
+  }
+
+  return Status::OK();
+}
+
+Status AdamWOptimizer::ComputeInternal(OpKernelContext* ctx) const {
+  const Tensor* learning_rate = ctx->Input<Tensor>(0);
+  const Tensor* step = ctx->Input<Tensor>(1);
+  const TensorSeq* weights = ctx->Input<TensorSeq>(2);
+  const TensorSeq* gradients = ctx->Input<TensorSeq>(3);
+  const TensorSeq* momentums_1 = ctx->Input<TensorSeq>(4);
+  const TensorSeq* momentums_2 = ctx->Input<TensorSeq>(5);
+
+  size_t num_of_weights = weights->Size();
+  size_t num_of_gradients = gradients->Size();
+  size_t num_of_momentums_1 = momentums_1->Size();
+  size_t num_of_momentums_2 = momentums_2->Size();
+
+  // Check the number of weights, gradients, momentums matchs.
+  ORT_ENFORCE(num_of_weights > 0, "invalid count of tensors in Seq<Tensor>.");
+  ORT_ENFORCE(num_of_weights == num_of_gradients, "number of weights and gradients mismatch.");
+  ORT_ENFORCE(num_of_gradients == num_of_momentums_1, "number of gradients and momentums_1 mismatch.");
+  ORT_ENFORCE(num_of_momentums_1 == num_of_momentums_2, "number of momentums_1 and momentums_2 mismatch.");
+
+  std::vector<int> tensor_sizes(num_of_weights);
+  std::vector<std::vector<void*>> grouped_tensor_pointers(num_of_weights);
+
+  for (size_t i = 0; i < num_of_weights; ++i) {
+    const Tensor& weight_tensor = weights->Get(i);
+    const Tensor& gradient_tensor = gradients->Get(i);
+    const Tensor& momentum_1_tensor = momentums_1->Get(i);
+    const Tensor& momentum_2_tensor = momentums_2->Get(i);
+
+    // Check the weight/gradient/momentums at the same index should have same shape.
+    ORT_ENFORCE(weight_tensor.Shape() == gradient_tensor.Shape(), "shape of weight and gradient mismatch.");
```

Comment:
good to also output the index/name of the weight for which it mismatched

Reply:
done